### ISSUE TITLE
Lack of comment in the doubled class causes "ParseError: syntax error, unexpected 'if' (T_IF), expecting function (T_FUNCTION) or const (T_CONST)"

### ISSUE THREAD

**Issue by eugenlisov** [Codeception/AspectMock#185]
Hi,

Love AspectMock! I finally managed to start testing stuff which I've previously found untestable.

However, I'm getting a very weird and annoying issue with it, though. I'm getting "ParseError: syntax error, unexpected 'if' (T_IF), expecting function (T_FUNCTION) or const (T_CONST)" without any apparent reason. 

![image](https://user-images.githubusercontent.com/9373003/89811845-ec30bf00-db47-11ea-80ae-fc234a784143.png)

I've done a bunch of tests to figure out what it could be. I've even went so far as to try to double a completely blank class, just it's definition.

Actually, I've now reproduced it:

Here's the test:
![image](https://user-images.githubusercontent.com/9373003/89812545-f4d5c500-db48-11ea-8a5b-8c1de41c5348.png)

Here's the Class
![image](https://user-images.githubusercontent.com/9373003/89812588-04550e00-db49-11ea-89e1-9940ef46167f.png)

And here's the Result::
![image](https://user-images.githubusercontent.com/9373003/89812618-12a32a00-db49-11ea-89cf-7d814f7649ef.png)

There's absolutely no 'if' in that class. It's as clean as it can be. 

At first I thought it was the bootstrap.php file, but it's all correct there:
![image](https://user-images.githubusercontent.com/9373003/89812822-6281f100-db49-11ea-882e-58288158db47.png)
 
The $kernel is initialized correctly.
![image](https://user-images.githubusercontent.com/9373003/89812890-775e8480-db49-11ea-83b9-f8929d9c13b8.png)

After a lot work trying to figure it out, I did find a quite easy fix, but which makes no sense.

Just by adding a simple comment "//"after opening the class, makes the test pass. Like this: 
![image](https://user-images.githubusercontent.com/9373003/89813089-cf958680-db49-11ea-8620-56a6b7b8319d.png)
This is the result: 
![image](https://user-images.githubusercontent.com/9373003/89813125-da501b80-db49-11ea-9166-002a4bef9b02.png)

It makes absolutely no sense to me why that fixes it.
I'm ok with that fix for now, as it allows me to focus on the rest of the project. It's not pretty having to add useless comments throughtout the code, but it does the job for now.

It would be nice to know what actually causes this issue and if there's anything more robust I can do, or can be added to AspectMock to avoid this in the future.

Thanks!

**Comment by dudapiotr** [comment]
Downgrade to older version  of `goaop/framework` helped me : "goaop/framework": "2.3.2". 
There must be some bug.

**Comment by eugenlisov** [comment]
Hi @dudapiotr ,

Thanks for the suggestion. I did the downgrade (from 2.3.4) in two different projects and  it worked 50%.

1. A small Laravel project. I did the downgrade and removed the comments. Tests run correctly: 
![image](https://user-images.githubusercontent.com/9373003/89920139-bac9f900-dc04-11ea-83b6-e2d6f0c2007e.png)

2. A WP plugin. I did the downgrade, removed the comments. But this time, I get the same errors: 
![image](https://user-images.githubusercontent.com/9373003/89920324-ee0c8800-dc04-11ea-82dc-9b99990b775b.png)
The class I'm doubling looks like this: 
![image](https://user-images.githubusercontent.com/9373003/89920458-1a280900-dc05-11ea-9b16-f90e37663823.png)
Once I put back the comment, the tests start running fine: 
![image](https://user-images.githubusercontent.com/9373003/89920602-4479c680-dc05-11ea-80c8-024f70cd5348.png)
The comment looks like this: 
![image](https://user-images.githubusercontent.com/9373003/89920691-5ce9e100-dc05-11ea-8efa-643aefbcfd62.png)

So, I'm guessing it's a combination between goaop/framework version and some other package that's not working right. But I'm really not capable of going though the composer.lock and trying to figure it out. There must be a better way than going line by line. I'll stick with the comments in the second project for now

**Comment by eugenlisov** [comment]
Found another thing that causes the same issue.
If I import more than one class at the top of the file it fails. If I leave just one, it works. 

So, for this I do another workaround: I inline the full namespace of each class. 
![image](https://user-images.githubusercontent.com/9373003/90031135-908c4000-dcc5-11ea-9d6c-3a48266d7909.png)
Not pretty, but it does the job.

**Comment by eugenlisov** [comment]
It turns out that whatever the bug is, it is causing way more problems than I initially thought. 
The most annoying was that it caused other tests that didn't use AspectMock to fail. For crazy reasons ( Unexpected "::"s where there were none, others that I can't remember). So I got to a point where half of the tests (the ones using AM) were working, the other failing. Then, I would disable AspectMock from bootstrap and the other half would start working and the AM ones fail. There was no way to do have a full suite test run from start to finish.

So I finally decided to remove AspectMock from the project and do my test doubles by extending the base classes and overwriting the methods that were calling in other dependencies. That was pretty easy, in my current project at least. 
I hope I don't get into the same issues with the other projects where I've started using it. I'm a bit nervous at the thought that I may go in too deep with AspectMock and then end up with tens of tests failing for no reason.. I'll keep an eye on that and if I get a hint of troubles, I'll have to switch to the test double trick I did above. How that won't be the case.

Thanks!

**Comment by thornview** [comment]
I'm seeing the same error and have been able to track it down a bit further, but don't know how to fix it. 

AspectMock overrides methods by adding an `if` statement to the beginning of the method.  For some reason, it is now adding an `if` statement to `switch` statements as well, which is not valid in PHP.  Here's a sample I captured using xdebug.  AspectMock adds the line `if (($__am_res = __amock_before($this, __CLASS__, __FUNCTION__, array(), false)) !== __AM_CONTINUE__) { return $__am_res; }` to the code.  In this sample you'll see AM has added it to the beginning of the method AND to the switch statement. 

```php   
 public function hasSignatureFile(): bool
    { if (($__am_res = __amock_before($this, __CLASS__, __FUNCTION__, array(), false)) !== __AM_CONTINUE__) { return $__am_res; }

      switch ($this->getRedcapPid()) { if (($__am_res = __amock_before($this, __CLASS__, __FUNCTION__, array(), false)) !== __AM_CONTINUE__) { return $__am_res; }

    case self::REDCAP_LC_HRPP:
            case self::REDCAP_LC_HRPP_CC:
            case self::REDCAP_LC_PI:
            case self::REDCAP_LC_PI_CC:
                return true;
            default:
                return false;
        }
    }
```

**Comment by sollidsnake** [comment]
This issue seems to be related to the `php-parser` package. I downgraded it to the version `4.6.0` and now AspectMock works with no issues.

**Comment by traceypooh** [comment]
THANK YOU @sollidsnake - this has been *plaguing* us for months now...
I found we needed to peg 1 hop further back, eg:

`"nikic/php-parser": "4.5.0",`

which, in turn, holds us back to phpunit v8.

But after days of trying to sort this out (for a newer, focal-based version of our monolith) finally getting non-corruptions for once!

**Comment by xfuturomax** [comment]
I have the same issue with `nikic/php-parser` version `> 4.5.0`. Some of files are generated with broken code.
For example when I run tests with `4.6`:
console/tests/_output/aopCache/common/lib/vendor/yiisoft/yii2/console/Application.php:108
is generated with broken code:
~~~
return require $file \Go\Instrument\Transformer\FilterInjectorTransformer::rewrite(;
                    , __DIR__)}
~~~
 in 4.5 the same string is:
~~~
return require \Go\Instrument\Transformer\FilterInjectorTransformer::rewrite($file, __DIR__);
~~~

Is it a problem of `nikic/php-parser` library or maybe `goaop/parser-reflection` because it requires this library? 

I want to upgrade to `phpunit 9` and it requires new version of this library:
~~~
phpunit/phpunit 9.3.0 requires phpunit/php-code-coverage ^9.0
phpunit/php-code-coverage 9.0.0 requires nikic/php-parser ^4.7
~~~

**Comment by xfuturomax** [comment]
I found linked issue: https://github.com/goaop/parser-reflection/issues/104

**Comment by xfuturomax** [comment]
I've done fix, but it is on review now.

If somebody want to test or use it on different environments before pull-request will be approved and new version will be released with this patch, I did forks of 3 repositories, you just need to specify it like this in your composer file:
~~~json
...
"repositories": [      
    {
      "type": "vcs",
      "url": "https://github.com/xfuturomax/framework"
    },
    {
      "type": "vcs",
      "url": "https://github.com/xfuturomax/parser-reflection"
    },
    {
      "type": "vcs",
      "url": "https://github.com/xfuturomax/AspectMock"
    }
  ],
...
"require-dev": {   
    "codeception/aspect-mock": "dev-fix/support-of-new-php-parser-composer"
  },
...
~~~

**Comment by traceypooh** [comment]
will try to give that a shot now @xfuturomax - thanks!

**Comment by traceypooh** [comment]
so far, i keep getting...

``` Problem 1
    - Installation request for codeception/aspect-mock dev-fix/support-of-new-php-parser-composer -> satisfiable by codeception/aspect-mock[dev-fix/support-of-new-php-parser-composer].
    - codeception/aspect-mock dev-fix/support-of-new-php-parser-composer requires goaop/framework dev-fix/support-of-new-php-parser-composer -> no matching package found.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.
```

**Comment by xfuturomax** [comment]
@traceypooh Did you specify repositories like in my example above?

**Comment by traceypooh** [comment]
i did.  but looks like (i'm a bit of a clowncar w/ keyboard keys salad over here with more than basic composer.json...)

but looks like finally got my setup to install like this:
```
...
    "require-dev": {
        "codeception/aspect-mock": "dev-fix/support-of-new-php-parser-composer",
        "goaop/framework": "dev-fix/support-of-new-php-parser-composer",
        "goaop/parser-reflection": "dev-fix/support-of-new-php-parser",
        "phpunit/phpunit": "*",
        "squizlabs/php_codesniffer": "3.5.6"
    }
...
```

**Comment by traceypooh** [comment]
*TESTS WORKING!*  🎉  🙏  🙇‍♀️

**Comment by troy-rudolph** [comment]
Hello, is there an E.T.A. for a release that contains this solution?  Thank you.

**Comment by troy-rudolph** [comment]
Hello, is there an E.T.A. for a release that contains this solution?  Thank you.

**Comment by Martin1982** [comment]
@Naktibalda it would be very much appreciated if these changes get released

**Comment by sollidsnake** [comment]
Looks like this package has been abandoned. Unless somebody forks it, I think this is it, but it would be nice to get a confirmation by the owner and maybe encourage new maintainers.

**Comment by Naktibalda** [comment]
@xfuturomax patch to goaop/framework was released in goaop/framework 3.0.0 version on 5th of April. It only supports PHP 7.4.

AspectMock v3 is not compatible with goaop/framework 3.0.0, I will make necessary changes and release AspectMock 4.0.0 supporting goaop/framework v3 only.

**Comment by Naktibalda** [comment]
This issue happens in AspectMock's own tests on [Github Actions ](https://github.com/Codeception/AspectMock/runs/2314193689) and it happened on Travis last year.

On Github Actions, error happens with PHP 7.0, 7.3 and 7.4, but it doesn't happen with 7.1 and 7.2
I tried to reproduce it locally, but it didn't happen on any PHP versions I had, I tried PHP 7.4, 7.3, 7.2 and 7.0.
I tried to use older versions of nikic/php-parser, goap/framework and  goaop/parser-reflection, but they all worked correctly.

Does anyone know what the issue is and how to fix it? Would adding version constraint for some dependency really help to avoid id?

**Comment by traceypooh** [comment]
the only fix i've figured out and we're currently using is (above) at 

https://github.com/Codeception/AspectMock/issues/185#issuecomment-699193577

we have 2 composer.json files -- one for xenial/php v7.0;  one for focal/php v7.4

```
cat composer.xenial.json 
{
    "require": {
        "braintree/braintree_php": "3.40.0",
        "cbschuld/browser.php": "1.9.1",
        "ezyang/htmlpurifier": "4.5.0",
        "ksubileau/color-thief-php": "1.4.1",
        "twig/twig": "1.15.1",
        "ua-parser/uap-php": "3.8.3",
        "google/apiclient": "2.8.3"
    },
    "require-dev": {
        "codeception/aspect-mock": "3.0.2",
        "nikic/php-parser": "4.5.0",
        "phpunit/phpunit": "6.5.14",
        "squizlabs/php_codesniffer": "*"
    }
}



cat composer.json
{
    "repositories": [
        {
          "type": "vcs",
          "url": "https://traceypooh:xxx@github.com/xfuturomax/framework"
        },
        {
          "type": "vcs",
          "url": "https://traceypooh:xxx@github.com/xfuturomax/parser-reflection"
        },
        {
          "type": "vcs",
          "url": "https://traceypooh:xxx@github.com/xfuturomax/AspectMock"
        }
      ],
    "require": {
        "braintree/braintree_php": "3.40.0",
        "cbschuld/browser.php": "1.9.1",
        "ezyang/htmlpurifier": "4.13",
        "ksubileau/color-thief-php": "1.4.1",
        "twig/twig": "1.15.1",
        "ua-parser/uap-php": "3.8.3",
        "google/apiclient": "2.8.3"
    },
    "require-dev": {
        "codeception/aspect-mock": "dev-fix/support-of-new-php-parser-composer",
        "goaop/framework": "dev-fix/support-of-new-php-parser-composer",
        "goaop/parser-reflection": "dev-fix/support-of-new-php-parser",
        "phpunit/phpunit": "*",
        "squizlabs/php_codesniffer": "*",
        "dealerdirect/phpcodesniffer-composer-installer": "^0.7.1",
        "sirbrillig/phpcs-variable-analysis": "^2.10"
    }
}

```

**Comment by traceypooh** [comment]
(and xxx is not my password ;) )

**Comment by traceypooh** [comment]
the (best i can tell) character-by-character `php-parser` parser is broken badly in php v7.4 and mangles files copied into the aspectmock tmp cache dir.  so aspectmock _pretty much doesnt work_ in php v7.4

**Comment by Naktibalda** [comment]
I created goaop-framework-v3 branch, could you test it on PHP 7.4?

https://github.com/Codeception/AspectMock/pull/193

**Comment by Naktibalda** [comment]
This issue happened on PHP 7.4 with goaop/framework v2: https://github.com/Codeception/AspectMock/runs/2314193689?check_suite_focus=true

but it doesn't happen with goaop/framework v3, so I consider it fixed. https://github.com/Codeception/AspectMock/pull/193/checks?check_run_id=2364951832

**Comment by Naktibalda** [comment]
I released it as 4.0.0 version, please upgrade.

https://github.com/Codeception/AspectMock/releases/tag/4.0.0

**Comment by traceypooh** [comment]
working *beautifully* so far for us over here, so many thanks! 🙏
